server/api/src/rsDataObjUnlink.cpp

```diff
@@ -34,12 +34,58 @@
 #include "rsRegDataObj.hpp"
 #include "rcMisc.h"
 
-// =-=-=-=-=-=-=-
 #include "irods_resource_backport.hpp"
 #include "irods_resource_redirect.hpp"
 #include "irods_hierarchy_parser.hpp"
 #include "irods_at_scope_exit.hpp"
 
+#define RODS_SERVER
+#include "irods_query.hpp"
+#undef RODS_SERVER
+
+#include "boost/filesystem/path.hpp"
+
+#include <string>
+#include <string_view>
+#include <iterator>
+
+namespace
+{
+    auto is_data_object_in_vault(rsComm_t& comm,
+                                 const boost::filesystem::path& logical_path,
+                                 std::string_view resc_hier) -> std::tuple<int, bool>
+    {
+        std::string vault_path;
+        if (const auto err = irods::get_vault_path_for_hier_string(resc_hier.data(), vault_path); !err.ok()) {
+            return {err.code(), false};
+        }
+
+        namespace fs = boost::filesystem;
+
+        std::string gql = "select DATA_PATH where COLL_NAME = '";
+        gql += logical_path.parent_path().c_str();
+        gql += "' and DATA_NAME = '";
+        gql += logical_path.filename().c_str();
+        gql += '\'';
+
+        fs::path child;
+        for (auto&& row : irods::query{&comm, gql}) {
+            child = row[0];
+        }
+
+        const fs::path parent = vault_path;
+
+        auto p_iter = std::begin(parent);
+        auto p_last = std::end(parent);
+        auto c_iter = std::begin(child);
+        auto c_last = std::end(child);
+
+        for (; p_iter != p_last && c_iter != c_last && *p_iter == *c_iter; ++p_iter, ++c_iter);
+
+        return {0, p_iter == p_last};
+    }
```

Comment:
Just making a note that server/core/include/physPath.hpp define `isInVault` which ostensibly performs the same task. This seems like a nice replacement. Perhaps a future improvement.

Reply:
Lets make a new issue for this one.

Reply:
I reviewed the version you mentioned and I agree the previous version should be replaced.

The `isInVault` implementation actually has a substring bug.